herddb-core/src/test/java/herddb/core/UpdateTest.java

```diff
@@ -183,4 +186,43 @@ public void updateMultiRowsWithValidationError() throws Exception {
         }
 
     }
+
+     @Test
+    public void upsertTest() throws Exception {
+        String nodeId = "localhost";
+        try (DBManager manager = new DBManager("localhost", new MemoryMetadataStorageManager(), new MemoryDataStorageManager(), new MemoryCommitLogManager(), null, null)) {
+            manager.start();
+            CreateTableSpaceStatement st1 = new CreateTableSpaceStatement("tblspace1", Collections.singleton(nodeId), nodeId, 1, 0, 0);
+            manager.executeStatement(st1, StatementEvaluationContext.DEFAULT_EVALUATION_CONTEXT(), TransactionContext.NO_TRANSACTION);
+            manager.waitForTablespace("tblspace1", 10000);
+
+            execute(manager, "CREATE TABLE tblspace1.tsql (k1 string primary key,n1 int,s1 string)", Collections.emptyList());
+
+            assertEquals(1, executeUpdate(manager, "INSERT INTO tblspace1.tsql(k1,n1) values(?,?)", Arrays.asList("mykey", Integer.valueOf(1234))).getUpdateCount());
+            try (DataScanner scan = scan(manager, "SELECT n1 from tblspace1.tsql where k1=?", Arrays.asList("mykey"))) {
+                List<DataAccessor> recordSet = scan.consumeAndClose();
+                assertEquals(1, recordSet.size());
+                assertEquals(1234, recordSet.get(0).get(0));
+            }
+            assertEquals(1, executeUpdate(manager, "UPSERT INTO tblspace1.tsql(k1,n1) values(?,?)", Arrays.asList("mykey", Integer.valueOf(1235))).getUpdateCount());
+            try (DataScanner scan = scan(manager, "SELECT n1 from tblspace1.tsql where k1=?", Arrays.asList("mykey"))) {
+                List<DataAccessor> recordSet = scan.consumeAndClose();
+                assertEquals(1, recordSet.size());
+                assertEquals(1235, recordSet.get(0).get(0));
+            }
+            assertEquals(4, executeUpdate(manager, "UPSERT INTO tblspace1.tsql(k1,n1)"
+                    + "values(?,?),(?,?),(?,?),(?,?)", Arrays.asList(
+                            "mykey", Integer.valueOf(1235),
+                            "mykey", Integer.valueOf(1236),
+                            "mykey", Integer.valueOf(1237),
+                            "mykey", Integer.valueOf(1238)
+                            )).getUpdateCount());
+            try (DataScanner scan = scan(manager, "SELECT n1 from tblspace1.tsql where k1=?", Arrays.asList("mykey"))) {
```

Comment:
If you call upsert on a table with non-null fields will it throw an exception regarding non null values? And if null values are allowed it will insert whatever is provided?

Reply:
It works exactly like an INSERT, falling back to a FULL record UPDATE.
So handling of null values will behave as usual.

Do you think it is worth to add an explicit test case?

Reply:
@amitvc  I have added more tests cases.